{
 "metadata": {
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7-final"
  },
  "orig_nbformat": 2,
  "kernelspec": {
   "name": "python3",
   "display_name": "Python 3",
   "language": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2,
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "datasetfname = os.path.join(\"input\",\"datasets.xlsx\")\n",
    "datapath = os.path.join(\"central\",\"groups\",\"murray-biocircuits\",\"ashur\",\"nanopore\")\n",
    "\n",
    "readsname = \"allreads.fastq\"\n",
    "outname = \"simprec.py\"\n",
    "\n",
    "df_data = pd.read_excel(datasetfname,sheet_name=\"alldata\")\n",
    "df_inducers = pd.read_excel(datasetfname,sheet_name=\"inducers\",header=13)\n",
    "df_seqs = pd.read_excel(datasetfname,sheet_name=\"sequences\")\n",
    "\n",
    "#layout\n",
    "#            plasmid                    genome        \n",
    "#================================---------------------------\n",
    "#   [bc>  [u21r><attB]<plasbc]<attL]   <uintf] <rc(bc)]\n",
    "#<barcode><prefixseq><variable_seq><postfixseq><rc(barcode)>\n",
    "#<barcode><newplasmidend1><plasbc1><genomechunk><rc(barcode)>\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "    date_sequenced  barcode genome int_control plasmid1 plasmid2 primerf  \\\n",
       "0           180201   L10R49  B_gen         s22    Bcsoo      NaN   UintF   \n",
       "1           180201  L10R126  B_gen         s22    Bcsoo      NaN   UintF   \n",
       "2           180201   L10L10  B_gen         s22    Bcsoo      NaN   UintF   \n",
       "3           180201   R49R49  B_gen         s22    Bcsoo      NaN   UintF   \n",
       "4           180201   L10R49  B_gen         s22    Bpcpc      NaN   UintF   \n",
       "..             ...      ...    ...         ...      ...      ...     ...   \n",
       "177         201212    npbc6    NaN       REC49      i70    i71-2   UintF   \n",
       "178         201212    npbc7    NaN       REC49      i70    i71-2   UintF   \n",
       "179         201212    npbc8    NaN       REC49      i70    i71-2   UintF   \n",
       "180         201212    npbc9    NaN       REC49      i70    i71-2   UintF   \n",
       "181         201212   npbc10    NaN       REC49      i70    i71-2   UintF   \n",
       "\n",
       "    primerr                c1                 c2   c3   c4        see  \\\n",
       "0       NaN         25,[atc2]                NaN  NaN  NaN        NaT   \n",
       "1       NaN         25,[atc1]                NaN  NaN  NaN        NaT   \n",
       "2       NaN          25,[atc]                NaN  NaN  NaN        NaT   \n",
       "3       NaN             25,[]                NaN  NaN  NaN        NaT   \n",
       "4       NaN         25,[atc2]                NaN  NaN  NaN        NaT   \n",
       "..      ...               ...                ...  ...  ...        ...   \n",
       "177  chlorR          5,[sal1]          25,[sal1]  NaN  NaN 2020-12-11   \n",
       "178  chlorR          5,[sal1]  25,[sal1,atc,cin]  NaN  NaN 2020-12-11   \n",
       "179  chlorR  5,[sal1,atc,ara]  25,[sal1,atc,cin]  NaN  NaN 2020-12-11   \n",
       "180  chlorR          5,[sal1]  25,[sal1,atc,ara]  NaN  NaN 2020-12-11   \n",
       "181  chlorR  5,[sal1,atc,cin]  25,[sal1,atc,ara]  NaN  NaN 2020-12-11   \n",
       "\n",
       "                          prefix variable1 variable2       suffix note  \n",
       "0                            NaN       NaN       NaN          NaN  NaN  \n",
       "1                            NaN       NaN       NaN          NaN  NaN  \n",
       "2                            NaN       NaN       NaN          NaN  NaN  \n",
       "3                            NaN       NaN       NaN          NaN  NaN  \n",
       "4                            NaN       NaN       NaN          NaN  NaN  \n",
       "..                           ...       ...       ...          ...  ...  \n",
       "177  [newplasmend2,newplasmend1]   plasbc1   plasbc2  genomechunk  NaN  \n",
       "178  [newplasmend2,newplasmend1]   plasbc1   plasbc2  genomechunk  NaN  \n",
       "179  [newplasmend2,newplasmend1]   plasbc1   plasbc2  genomechunk  NaN  \n",
       "180  [newplasmend2,newplasmend1]   plasbc1   plasbc2  genomechunk  NaN  \n",
       "181  [newplasmend2,newplasmend1]   plasbc1   plasbc2  genomechunk  NaN  \n",
       "\n",
       "[182 rows x 18 columns]"
      ],
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>date_sequenced</th>\n      <th>barcode</th>\n      <th>genome</th>\n      <th>int_control</th>\n      <th>plasmid1</th>\n      <th>plasmid2</th>\n      <th>primerf</th>\n      <th>primerr</th>\n      <th>c1</th>\n      <th>c2</th>\n      <th>c3</th>\n      <th>c4</th>\n      <th>see</th>\n      <th>prefix</th>\n      <th>variable1</th>\n      <th>variable2</th>\n      <th>suffix</th>\n      <th>note</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>180201</td>\n      <td>L10R49</td>\n      <td>B_gen</td>\n      <td>s22</td>\n      <td>Bcsoo</td>\n      <td>NaN</td>\n      <td>UintF</td>\n      <td>NaN</td>\n      <td>25,[atc2]</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaT</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>180201</td>\n      <td>L10R126</td>\n      <td>B_gen</td>\n      <td>s22</td>\n      <td>Bcsoo</td>\n      <td>NaN</td>\n      <td>UintF</td>\n      <td>NaN</td>\n      <td>25,[atc1]</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaT</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>180201</td>\n      <td>L10L10</td>\n      <td>B_gen</td>\n      <td>s22</td>\n      <td>Bcsoo</td>\n      <td>NaN</td>\n      <td>UintF</td>\n      <td>NaN</td>\n      <td>25,[atc]</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaT</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>180201</td>\n      <td>R49R49</td>\n      <td>B_gen</td>\n      <td>s22</td>\n      <td>Bcsoo</td>\n      <td>NaN</td>\n      <td>UintF</td>\n      <td>NaN</td>\n      <td>25,[]</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaT</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>180201</td>\n      <td>L10R49</td>\n      <td>B_gen</td>\n      <td>s22</td>\n      <td>Bpcpc</td>\n      <td>NaN</td>\n      <td>UintF</td>\n      <td>NaN</td>\n      <td>25,[atc2]</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaT</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>...</th>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n    </tr>\n    <tr>\n      <th>177</th>\n      <td>201212</td>\n      <td>npbc6</td>\n      <td>NaN</td>\n      <td>REC49</td>\n      <td>i70</td>\n      <td>i71-2</td>\n      <td>UintF</td>\n      <td>chlorR</td>\n      <td>5,[sal1]</td>\n      <td>25,[sal1]</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>2020-12-11</td>\n      <td>[newplasmend2,newplasmend1]</td>\n      <td>plasbc1</td>\n      <td>plasbc2</td>\n      <td>genomechunk</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>178</th>\n      <td>201212</td>\n      <td>npbc7</td>\n      <td>NaN</td>\n      <td>REC49</td>\n      <td>i70</td>\n      <td>i71-2</td>\n      <td>UintF</td>\n      <td>chlorR</td>\n      <td>5,[sal1]</td>\n      <td>25,[sal1,atc,cin]</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>2020-12-11</td>\n      <td>[newplasmend2,newplasmend1]</td>\n      <td>plasbc1</td>\n      <td>plasbc2</td>\n      <td>genomechunk</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>179</th>\n      <td>201212</td>\n      <td>npbc8</td>\n      <td>NaN</td>\n      <td>REC49</td>\n      <td>i70</td>\n      <td>i71-2</td>\n      <td>UintF</td>\n      <td>chlorR</td>\n      <td>5,[sal1,atc,ara]</td>\n      <td>25,[sal1,atc,cin]</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>2020-12-11</td>\n      <td>[newplasmend2,newplasmend1]</td>\n      <td>plasbc1</td>\n      <td>plasbc2</td>\n      <td>genomechunk</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>180</th>\n      <td>201212</td>\n      <td>npbc9</td>\n      <td>NaN</td>\n      <td>REC49</td>\n      <td>i70</td>\n      <td>i71-2</td>\n      <td>UintF</td>\n      <td>chlorR</td>\n      <td>5,[sal1]</td>\n      <td>25,[sal1,atc,ara]</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>2020-12-11</td>\n      <td>[newplasmend2,newplasmend1]</td>\n      <td>plasbc1</td>\n      <td>plasbc2</td>\n      <td>genomechunk</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>181</th>\n      <td>201212</td>\n      <td>npbc10</td>\n      <td>NaN</td>\n      <td>REC49</td>\n      <td>i70</td>\n      <td>i71-2</td>\n      <td>UintF</td>\n      <td>chlorR</td>\n      <td>5,[sal1,atc,cin]</td>\n      <td>25,[sal1,atc,ara]</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>2020-12-11</td>\n      <td>[newplasmend2,newplasmend1]</td>\n      <td>plasbc1</td>\n      <td>plasbc2</td>\n      <td>genomechunk</td>\n      <td>NaN</td>\n    </tr>\n  </tbody>\n</table>\n<p>182 rows Ã— 18 columns</p>\n</div>"
     },
     "metadata": {},
     "execution_count": 2
    }
   ],
   "source": [
    "df_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "date of seq is 181113\n",
      "get reads from central\\groups\\murray-biocircuits\\ashur\\nanopore\\181113\\allreads.fastq\n",
      "prefixseqslist is ['AGGTATGATCCTGACGACGGAGCACGCCGTCGTCGACAAGCC']\n",
      "plasbcs is [['CTGACAGCTAGCTCAGTCCTAGGTATAATGCTAGC', 'TTTCAATTTAATCATCCGGCTCGTATAATGTGTGGA']]\n",
      "postfixseq is ['CAAGCCCATTATTACCCTGTTATCCCTAGACACCAATCAGAGGCCACA']\n",
      "bcseqs is {'npbc1': 'AAGAAAGTTGTCGGTGTCTTTGT', 'npbc2': 'TCGATTCCGTTTGTAGTCGTCTG', 'npbc3': 'GAGTCTTGTGTCCCAGTTACCAG', 'npbc4': 'TTCGGATTCTATCGTGTTTCCCT', 'npbc5': 'CTTGTCCAGGGTTTGTGTAACCT', 'npbc7': 'GTGTTACCGTGGGAATGAATCCT', 'npbc8': 'TTCAGGGAACAAACCAAGTTACG', 'npbc9': 'AACTAGGCACAGCGAGTCTTGGT', 'npbc10': 'AAGCGTTGAAACCTTTGTCCTCT', 'npbc11': 'GTTTCATCTATCGGAGGGAATGG'}\n",
      "condnames is ['4,[sal1]_4,[sal1]_4,[sal1]_25,[sal1]', '4,[atc1]_4,[atc1]_4,[atc1]_25,[atc1]', '4,[atc1]_4,[atc1]_4,[atc1]_25,[atc1]', '4,[sal1,atc,ara]_4,[sal1,atc,ara]_4,[sal1,atc,ara]_25,[sal1,atc,cin]', '4,[sal,atc1,ara]_4,[sal,atc1,ara]_4,[sal,atc1,ara]_25,[sal,atc1,cin]', '4,[sal1]_4,[sal1]_4,[sal1]_25,[sal1]', '4,[atc1]_4,[atc1]_4,[atc1]_4,[atc1]', '4,[atc1]_4,[atc1]_4,[atc1]_4,[atc1]', '4,[sal1,atc,ara]_4,[sal1,atc,ara]_4,[sal1,atc,ara]_25,[sal1,atc,cin]', '4,[sal,atc1,ara]_4,[sal,atc1,ara]_4,[sal,atc1,ara]_25,[sal,atc1,cin]']\n",
      "put output into central\\groups\\murray-biocircuits\\ashur\\nanopore\\181113\\181113_simprec.py\n",
      "date of seq is 181203\n",
      "get reads from central\\groups\\murray-biocircuits\\ashur\\nanopore\\181203\\allreads.fastq\n",
      "prefixseqslist is ['AGGTATGATCCTGACGACGGAGCACGCCGTCGTCGACAAGCC']\n",
      "plasbcs is [['CTGACAGCTAGCTCAGTCCTAGGTATAATGCTAGC', 'TTTCAATTTAATCATCCGGCTCGTATAATGTGTGGA']]\n",
      "postfixseq is ['CAAGCCCATTATTACCCTGTTATCCCTAGACACCAATCAGAGGCCACA']\n",
      "bcseqs is {'npbc1': 'AAGAAAGTTGTCGGTGTCTTTGT', 'npbc2': 'TCGATTCCGTTTGTAGTCGTCTG', 'npbc3': 'GAGTCTTGTGTCCCAGTTACCAG', 'npbc4': 'TTCGGATTCTATCGTGTTTCCCT', 'npbc5': 'CTTGTCCAGGGTTTGTGTAACCT', 'npbc6': 'TTCTCGCAAAGGCAGAAAGTAGT', 'npbc7': 'GTGTTACCGTGGGAATGAATCCT', 'npbc8': 'TTCAGGGAACAAACCAAGTTACG', 'npbc9': 'AACTAGGCACAGCGAGTCTTGGT', 'npbc10': 'AAGCGTTGAAACCTTTGTCCTCT'}\n",
      "condnames is ['4,[sal1]_4,[sal1]_4,[sal1]_4,[sal1]', '4,[atc,ara,sal1]_4,[atc,ara,sal1]_4,[atc,ara,sal1]_4,[atc,ara,sal1]', '4,[atc,cin,sal1]_4,[atc,cin,sal1]_4,[atc,cin,sal1]_4,[atc,cin,sal1]', '4,[atc,ara,sal1]_4,[atc,ara,sal1]_4,[atc,cin,sal1]_4,[atc,cin,sal1]', '4,[atc,ara,sal1]_4,[atc,ara,sal1]_4,[atc,cin,sal1]_4,[atc,ara,sal1]', '4,[atc1]_4,[atc1]_4,[atc1]_4,[atc1]', '4,[atc1,ara,sal]_4,[atc1,ara,sal]_4,[atc1,ara,sal]_4,[atc1,ara,sal]', '4,[atc1,cin,sal]_4,[atc1,cin,sal]_4,[atc1,cin,sal]_4,[atc1,cin,sal]', '4,[atc1,ara,sal]_4,[atc1,ara,sal]_4,[atc1,cin,sal]_4,[atc1,cin,sal]', '4,[atc1,ara,sal]_4,[atc1,ara,sal]_4,[atc1,cin,sal]_4,[atc1,ara,sal]']\n",
      "put output into central\\groups\\murray-biocircuits\\ashur\\nanopore\\181203\\181203_simprec.py\n",
      "date of seq is 190118\n",
      "get reads from central\\groups\\murray-biocircuits\\ashur\\nanopore\\190118\\allreads.fastq\n",
      "prefixseqslist is ['AGGTATGATCCTGACGACGGAGCACGCCGTCGTCGACAAGCC']\n",
      "plasbcs is [['CTGACAGCTAGCTCAGTCCTAGGTATAATGCTAGC', 'TTTCAATTTAATCATCCGGCTCGTATAATGTGTGGA']]\n",
      "postfixseq is ['CAAGCCCATTATTACCCTGTTATCCCTAGACACCAATCAGAGGCCACA']\n",
      "bcseqs is {'npbc1': 'AAGAAAGTTGTCGGTGTCTTTGT', 'npbc2': 'TCGATTCCGTTTGTAGTCGTCTG', 'npbc3': 'GAGTCTTGTGTCCCAGTTACCAG', 'npbc4': 'TTCGGATTCTATCGTGTTTCCCT', 'npbc5': 'CTTGTCCAGGGTTTGTGTAACCT', 'npbc6': 'TTCTCGCAAAGGCAGAAAGTAGT', 'npbc7': 'GTGTTACCGTGGGAATGAATCCT', 'npbc8': 'TTCAGGGAACAAACCAAGTTACG', 'npbc9': 'AACTAGGCACAGCGAGTCTTGGT', 'npbc10': 'AAGCGTTGAAACCTTTGTCCTCT', 'npbc11': 'GTTTCATCTATCGGAGGGAATGG', 'npbc12': 'CAGGTAGAAAGAAGCAGAATCGG'}\n",
      "condnames is ['.5,[]_24.5,[]', '.5,[cin,sal]_24.5,[]', '1,[cin,sal]_24,[]', '3,[cin,sal_21,[]', '.5,[atc,ara,sal]_24.5,[atc,ara,sal]', '.5,[ara,sal]_24.5,[atc,ara,sal]', '1,[ara,sal]_24,[atc,ara,sal]', '3,[ara,sal]_21,[atc,ara,sal]', '.5,[atc,sal]_24.5,[atc,sal]', '.5,[sal]_24.5,[atc,sal]', '1,[sal]_24,[atc,sal]', '3,[sal]_21,[atc,sal]']\n",
      "put output into central\\groups\\murray-biocircuits\\ashur\\nanopore\\190118\\190118_simprec.py\n",
      "date of seq is 190207\n",
      "get reads from central\\groups\\murray-biocircuits\\ashur\\nanopore\\190207\\allreads.fastq\n",
      "prefixseqslist is ['AGGTATGATCCTGACGACGGAGCACGCCGTCGTCGACAAGCC']\n",
      "plasbcs is [['CTGACAGCTAGCTCAGTCCTAGGTATAATGCTAGC', 'TTTCAATTTAATCATCCGGCTCGTATAATGTGTGGA']]\n",
      "postfixseq is ['CAAGCCCATTATTACCCTGTTATCCCTAGACACCAATCAGAGGCCACA']\n",
      "bcseqs is {'npbc1': 'AAGAAAGTTGTCGGTGTCTTTGT', 'npbc2': 'TCGATTCCGTTTGTAGTCGTCTG', 'npbc3': 'GAGTCTTGTGTCCCAGTTACCAG', 'npbc4': 'TTCGGATTCTATCGTGTTTCCCT', 'npbc5': 'CTTGTCCAGGGTTTGTGTAACCT', 'npbc6': 'TTCTCGCAAAGGCAGAAAGTAGT', 'npbc7': 'GTGTTACCGTGGGAATGAATCCT', 'npbc8': 'TTCAGGGAACAAACCAAGTTACG', 'npbc9': 'AACTAGGCACAGCGAGTCTTGGT', 'npbc10': 'AAGCGTTGAAACCTTTGTCCTCT', 'npbc11': 'GTTTCATCTATCGGAGGGAATGG', 'npbc12': 'CAGGTAGAAAGAAGCAGAATCGG', 'npbc13': 'AGAACGACTTCCATACTCGTGTG', 'npbc14': 'AACGAGTCTCTTGGGACCCATAG', 'npbc15': 'AGGTCTACCTCGCTAACACCACT'}\n",
      "condnames is ['6,[atc3,sal1]_6,[atc3,ara,sal1]_42[,atc3,cin,sal1]', '6,[atc3,sal2]_6,[atc3,ara,sal2]_42[,atc3,cin,sal2]', '6,[atc3,sal3]_6,[atc3,ara,sal3]_42[,atc3,cin,sal3]', '6,[atc3,sal1]_6,[atc3,cin,sal1]_42,[atc3,ara,sal1]', '6,[atc3,sal2]_6,[atc3,cin,sal2]_42,[atc3,ara,sal2]', '6,[atc3,sal3]_6,[atc3,cin,sal3]_42,[atc3,ara,sal3]', '6,[atc3,sal1]_6,[atc3,sal1]_42,[atc3,sal1]', '6,[atc3,sal2]_6,[atc3,sal2]_42,[atc3,sal2]', '6,[atc3,sal3]_6,[atc3,sal3]_42,[atc3,sal3]', '6,[sal3]_6,[sal3]_42,[sal3,cin]', '6,[sal3,atc]_6,[sal3,ara,atc]_42,[sal3,cin,atc]', '6,[sal3]_6,[sal3,cin]_42,[sal3]', '6,[sal3,atc]_6,[sal3,cin,atc]_42,[sal3,ara,atc]', '6,[sal3]_6,[sal3]_42,[sal3]', '6,[sal3,atc]_6,[sal3,atc]_42,[sal3,atc]']\n",
      "put output into central\\groups\\murray-biocircuits\\ashur\\nanopore\\190207\\190207_simprec.py\n",
      "date of seq is 190220\n",
      "get reads from central\\groups\\murray-biocircuits\\ashur\\nanopore\\190220\\allreads.fastq\n",
      "prefixseqslist is ['AGGTATGATCCTGACGACGGAGCACGCCGTCGTCGACAAGCC']\n",
      "plasbcs is [['CTGACAGCTAGCTCAGTCCTAGGTATAATGCTAGC', 'TTTCAATTTAATCATCCGGCTCGTATAATGTGTGGA']]\n",
      "postfixseq is ['CAAGCCCATTATTACCCTGTTATCCCTAGACACCAATCAGAGGCCACA']\n",
      "bcseqs is {'npbc3': 'GAGTCTTGTGTCCCAGTTACCAG', 'npbc4': 'TTCGGATTCTATCGTGTTTCCCT', 'npbc5': 'CTTGTCCAGGGTTTGTGTAACCT', 'npbc8': 'TTCAGGGAACAAACCAAGTTACG', 'npbc9': 'AACTAGGCACAGCGAGTCTTGGT', 'npbc10': 'AAGCGTTGAAACCTTTGTCCTCT', 'npbc13': 'AGAACGACTTCCATACTCGTGTG', 'npbc14': 'AACGAGTCTCTTGGGACCCATAG', 'npbc15': 'AGGTCTACCTCGCTAACACCACT'}\n",
      "condnames is ['6,[atc3,sal1]_6,[atc3,ara,sal1]_42[,atc3,cin,sal1]', '6,[atc3,sal2]_6,[atc3,ara,sal2]_42[,atc3,cin,sal2]', '6,[atc3,sal3]_6,[atc3,ara,sal3]_42[,atc3,cin,sal3]', '6,[atc3,sal1]_6,[atc3,cin,sal1]_42,[atc3,ara,sal1]', '6,[atc3,sal2]_6,[atc3,cin,sal2]_42,[atc3,ara,sal2]', '6,[atc3,sal3]_6,[atc3,cin,sal3]_42,[atc3,ara,sal3]', '6,[atc3,sal1]_6,[atc3,sal1]_42,[atc3,sal1]', '6,[atc3,sal2]_6,[atc3,sal2]_42,[atc3,sal2]', '6,[atc3,sal3]_6,[atc3,sal3]_42,[atc3,sal3]']\n",
      "put output into central\\groups\\murray-biocircuits\\ashur\\nanopore\\190220\\190220_simprec.py\n",
      "date of seq is 190228\n",
      "get reads from central\\groups\\murray-biocircuits\\ashur\\nanopore\\190228\\allreads.fastq\n",
      "prefixseqslist is ['AGGTATGATCCTGACGACGGAGCACGCCGTCGTCGACAAGCC']\n",
      "plasbcs is [['CTGACAGCTAGCTCAGTCCTAGGTATAATGCTAGC', 'TTTCAATTTAATCATCCGGCTCGTATAATGTGTGGA']]\n",
      "postfixseq is ['CAAGCCCATTATTACCCTGTTATCCCTAGACACCAATCAGAGGCCACA']\n",
      "bcseqs is {'npbc7': 'GTGTTACCGTGGGAATGAATCCT', 'npbc8': 'TTCAGGGAACAAACCAAGTTACG', 'npbc9': 'AACTAGGCACAGCGAGTCTTGGT', 'npbc10': 'AAGCGTTGAAACCTTTGTCCTCT', 'npbc11': 'GTTTCATCTATCGGAGGGAATGG', 'npbc12': 'CAGGTAGAAAGAAGCAGAATCGG', 'npbc13': 'AGAACGACTTCCATACTCGTGTG', 'npbc14': 'AACGAGTCTCTTGGGACCCATAG', 'npbc15': 'AGGTCTACCTCGCTAACACCACT'}\n",
      "condnames is ['7.2,[sal1]_0.16,[sal1,atc]_7.2,[sal1]_0.16,[sal1,atc]', '7.2,[sal1,ara]_0.16,[sal1,ara,atc]_7.2,[sal1,cin]_0.16,[sal1,cin,atc]', '7.2,[sal1,cin]_0.16,[sal1,cin,atc]_7.2,[sal1,ara]_0.16,[sal1,ara,atc]', '6.5,[sal1]_0.86,[sal1,atc]_6.5,[sal1]_0.86,[sal1,atc]', '6.5,[sal1,ara]_0.86,[sal1,ara,atc]_6.5,[sal1,cin]_0.86,[sal1,cin,atc]', '6.5,[sal1,cin]_0.86,[sal1,cin,atc]_6.5,[sal1,ara]_0.86,[sal1,ara,atc]', '5,[atc1]_2.36,[atc1,sal]_5,[atc1]_2.36,[atc1,sal]', '5,[atc1,ara]_2.36,[atc1,ara,sal]_5,[atc1,cin]_2.36,[atc1,cin,sal]', '5,[atc1,cin]_2.36,[atc1,cin,sal]_5,[atc1,ara]_2.36,[atc1,ara,sal]']\n",
      "put output into central\\groups\\murray-biocircuits\\ashur\\nanopore\\190228\\190228_simprec.py\n",
      "date of seq is 190308\n",
      "get reads from central\\groups\\murray-biocircuits\\ashur\\nanopore\\190308\\allreads.fastq\n",
      "prefixseqslist is ['AGGTATGATCCTGACGACGGAGCACGCCGTCGTCGACAAGCC']\n",
      "plasbcs is [['CTGACAGCTAGCTCAGTCCTAGGTATAATGCTAGC', 'TTTCAATTTAATCATCCGGCTCGTATAATGTGTGGA']]\n",
      "postfixseq is ['CAAGCCCATTATTACCCTGTTATCCCTAGACACCAATCAGAGGCCACA']\n",
      "bcseqs is {'npbc1': 'AAGAAAGTTGTCGGTGTCTTTGT', 'npbc2': 'TCGATTCCGTTTGTAGTCGTCTG', 'npbc3': 'GAGTCTTGTGTCCCAGTTACCAG', 'npbc4': 'TTCGGATTCTATCGTGTTTCCCT', 'npbc5': 'CTTGTCCAGGGTTTGTGTAACCT', 'npbc6': 'TTCTCGCAAAGGCAGAAAGTAGT', 'npbc7': 'GTGTTACCGTGGGAATGAATCCT', 'npbc8': 'TTCAGGGAACAAACCAAGTTACG', 'npbc9': 'AACTAGGCACAGCGAGTCTTGGT', 'npbc10': 'AAGCGTTGAAACCTTTGTCCTCT', 'npbc11': 'GTTTCATCTATCGGAGGGAATGG', 'npbc12': 'CAGGTAGAAAGAAGCAGAATCGG', 'npbc13': 'AGAACGACTTCCATACTCGTGTG', 'npbc14': 'AACGAGTCTCTTGGGACCCATAG', 'npbc15': 'AGGTCTACCTCGCTAACACCACT'}\n",
      "condnames is ['4,[sal]_1,[sal,atc]_4,[sal]_1,[sal,atc]', '4.25,[sal]_0.75,[sal,atc]_4.25,[sal]_0.75,[sal,atc]', '4.5,[sal]_0.5,[sal,atc]_4.5,[sal]_0.5,[sal,atc]', '4.83,[sal]_0.17,[sal,atc]_4.83,[sal]_0.17,[sal,atc]', '4.92,[sal]_0.08,[sal,atc]_4.92,[sal]_0.08,[sal,atc]', '4,[ara,sal]_1,[ara,sal,atc]_4,[cin,sal]_1,[cin,sal,atc]', '4.25,[ara,sal]_0.75,[ara,sal,atc]_4.25,[cin,sal]_0.75,[cin,sal,atc]', '4.5,[ara,sal]_0.5,[ara,sal,atc]_4.5,[cin,sal]_0.5,[cin,sal,atc]', '4.83,[ara,sal]_0.17,[ara,sal,atc]_4.83,[cin,sal]_0.17,[cin,sal,atc]', '4.92,[ara,sal]_0.08,[ara,sal,atc]_4.92,[cin,sal]_0.08,[cin,sal,atc]', '4,[cin,sal]_1,[cin,sal,atc]_4,[ara,sal]_1,[ara,sal,atc]', '4.25,[cin,sal]_0.75,[cin,sal,atc]_4.25,[ara,sal]_0.75,[ara,sal,atc]', '4.5,[cin,sal]_0.5,[cin,sal,atc]_4.5,[ara,sal]_0.5,[ara,sal,atc]', '4.83,[cin,sal]_0.17,[cin,sal,atc]_4.83,[ara,sal]_0.17,[ara,sal,atc]', '4.92,[cin,sal]_0.08,[cin,sal,atc]_4.92,[ara,sal]_0.08,[ara,sal,atc]']\n",
      "put output into central\\groups\\murray-biocircuits\\ashur\\nanopore\\190308\\190308_simprec.py\n",
      "date of seq is 190428_1\n",
      "get reads from central\\groups\\murray-biocircuits\\ashur\\nanopore\\190428_1\\allreads.fastq\n",
      "prefixseqslist is ['AGGTATGATCCTGACGACGGAGCACGCCGTCGTCGACAAGCC']\n",
      "plasbcs is [['CTGACAGCTAGCTCAGTCCTAGGTATAATGCTAGC', 'TTTCAATTTAATCATCCGGCTCGTATAATGTGTGGA']]\n",
      "postfixseq is ['CAAGCCCATTATTACCCTGTTATCCCTAGACACCAATCAGAGGCCACA']\n",
      "bcseqs is {'npbc1': 'AAGAAAGTTGTCGGTGTCTTTGT', 'npbc2': 'TCGATTCCGTTTGTAGTCGTCTG', 'npbc3': 'GAGTCTTGTGTCCCAGTTACCAG', 'npbc4': 'TTCGGATTCTATCGTGTTTCCCT', 'npbc5': 'CTTGTCCAGGGTTTGTGTAACCT', 'npbc6': 'TTCTCGCAAAGGCAGAAAGTAGT', 'npbc7': 'GTGTTACCGTGGGAATGAATCCT', 'npbc8': 'TTCAGGGAACAAACCAAGTTACG', 'npbc9': 'AACTAGGCACAGCGAGTCTTGGT', 'npbc10': 'AAGCGTTGAAACCTTTGTCCTCT', 'npbc11': 'GTTTCATCTATCGGAGGGAATGG', 'npbc12': 'CAGGTAGAAAGAAGCAGAATCGG', 'npbc13': 'AGAACGACTTCCATACTCGTGTG'}\n",
      "condnames is ['4,[]_27,[sal2]', '4,[atc]_27,[atc,sal2]', '4,[atc1]_27,[atc1,sal2]', '4,[atc3]_27,[atc3,sal2]', '4,[atc4]_27,[atc4,sal2]', '4,[atc,ara]_27,[atc,ara,sal2]', '4,[atc1,ara]_27,[atc1,ara,sal2]', '4,[atc3,ara]_27,[atc3,ara,sal2]', '4,[atc4,ara]_27,[atc4,ara,sal2]', '4,[atc,cin]_27,[atc,cin,sal2]', '4,[atc1,cin]_27,[atc1,cin,sal2]', '4,[atc3,cin]_27,[atc3,cin,sal2]', '4,[atc4,cin]_27,[atc4,cin,sal2]']\n",
      "put output into central\\groups\\murray-biocircuits\\ashur\\nanopore\\190428_1\\190428_1_simprec.py\n",
      "date of seq is 190428_2\n",
      "get reads from central\\groups\\murray-biocircuits\\ashur\\nanopore\\190428_2\\allreads.fastq\n",
      "prefixseqslist is ['AGGTATGATCCTGACGACGGAGCACGCCGTCGTCGACAAGCC']\n",
      "plasbcs is [['CTGACAGCTAGCTCAGTCCTAGGTATAATGCTAGC', 'TTTCAATTTAATCATCCGGCTCGTATAATGTGTGGA']]\n",
      "postfixseq is ['CAAGCCCATTATTACCCTGTTATCCCTAGACACCAATCAGAGGCCACA']\n",
      "bcseqs is {'npbc1': 'AAGAAAGTTGTCGGTGTCTTTGT', 'npbc2': 'TCGATTCCGTTTGTAGTCGTCTG', 'npbc3': 'GAGTCTTGTGTCCCAGTTACCAG', 'npbc4': 'TTCGGATTCTATCGTGTTTCCCT', 'npbc5': 'CTTGTCCAGGGTTTGTGTAACCT', 'npbc6': 'TTCTCGCAAAGGCAGAAAGTAGT', 'npbc7': 'GTGTTACCGTGGGAATGAATCCT', 'npbc8': 'TTCAGGGAACAAACCAAGTTACG', 'npbc9': 'AACTAGGCACAGCGAGTCTTGGT', 'npbc10': 'AAGCGTTGAAACCTTTGTCCTCT', 'npbc11': 'GTTTCATCTATCGGAGGGAATGG', 'npbc12': 'CAGGTAGAAAGAAGCAGAATCGG', 'npbc13': 'AGAACGACTTCCATACTCGTGTG'}\n",
      "condnames is ['4,[]_27,[atc2]', '4,[sal]_27,[sal,atc2]', '4,[sal1]_27,[sal1,atc2]', '4,[sal3]_27,[sal3,atc2]', '4,[sal4]_27,[sal4,atc2]', '4,[sal,ara]_27,[sal,ara,atc2]', '4,[sal1,ara]_27,[sal1,ara,atc2]', '4,[sal3,ara]_27,[sal3,ara,atc2]', '4,[sal4,ara]_27,[sal4,ara,atc2]', '4,[sal,cin]_27,[sal,cin,atc2]', '4,[sal1,cin]_27,[sal1,cin,atc2]', '4,[sal3,cin]_27,[sal3,cin,atc2]', '4,[sal4,cin]_27,[sal4,cin,atc2]']\n",
      "put output into central\\groups\\murray-biocircuits\\ashur\\nanopore\\190428_2\\190428_2_simprec.py\n",
      "date of seq is 191011\n",
      "get reads from central\\groups\\murray-biocircuits\\ashur\\nanopore\\191011\\allreads.fastq\n",
      "prefixseqslist is ['AGGTATGATCCTGACGACGGAGCACGCCGTCGTCGACAAGCC']\n",
      "plasbcs is [['CCGGTTCGTTCCAAGCAACTAGTTTATCATCTAACTAG', 'CTCACCAGCTGATGAGTAACTATTCCTGCTCCATTGGT']]\n",
      "postfixseq is ['CAAGCCCATTATTACCCTGTTATCCCTAGACACCAATCAGAGGCCACA']\n",
      "bcseqs is {'npbc1': 'AAGAAAGTTGTCGGTGTCTTTGT', 'npbc2': 'TCGATTCCGTTTGTAGTCGTCTG', 'npbc3': 'GAGTCTTGTGTCCCAGTTACCAG', 'npbc4': 'TTCGGATTCTATCGTGTTTCCCT', 'npbc5': 'CTTGTCCAGGGTTTGTGTAACCT', 'npbc6': 'TTCTCGCAAAGGCAGAAAGTAGT', 'npbc7': 'GTGTTACCGTGGGAATGAATCCT', 'npbc8': 'TTCAGGGAACAAACCAAGTTACG', 'npbc9': 'AACTAGGCACAGCGAGTCTTGGT', 'npbc10': 'AAGCGTTGAAACCTTTGTCCTCT', 'npbc11': 'GTTTCATCTATCGGAGGGAATGG', 'npbc12': 'CAGGTAGAAAGAAGCAGAATCGG'}\n",
      "condnames is ['9,[]_9,[sal]_48,[sal]', '9,[]_9,[sal1]_48,[sal1]', '9,[]_9,[sal2]_48,[sal2]', '9,[]_9,[sal]_48,[sal]', '9,[]_9,[sal1]_48,[sal1]', '9,[]_9,[sal2]_48,[sal2]', '9,[atc]_9,[atc,sal2]_48,[atc,sal2]', '9,[atc,cin]_9,[atc,cin,sal2]_48,[atc,cin,sal2]', '9,[atc,ara]_9,[atc,ara,sal2]_48,[atc,ara,sal2]', '9,[atc]_9,[atc,sal2]_48,[atc,sal2]', '9,[atc,cin]_9,[atc,cin,sal2]_48,[atc,cin,sal2]', '9,[atc,ara]_9,[atc,ara,sal2]_48,[atc,ara,sal2]']\n",
      "put output into central\\groups\\murray-biocircuits\\ashur\\nanopore\\191011\\191011_simprec.py\n",
      "date of seq is 191114\n",
      "get reads from central\\groups\\murray-biocircuits\\ashur\\nanopore\\191114\\allreads.fastq\n",
      "prefixseqslist is ['AGGTATGATCCTGACGACGGAGCACGCCGTCGTCGACAAGCC']\n",
      "plasbcs is [['CCGGTTCGTTCCAAGCAACTAGTTTATCATCTAACTAG'], ['CTCACCAGCTGATGAGTAACTATTCCTGCTCCATTGGT', 'CTCACCAGCTGATGAGTAACTATTCCTGCTCCATTGGT']]\n",
      "postfixseq is ['CAAGCCCATTATTACCCTGTTATCCCTAGACACCAATCAGAGGCCACA']\n",
      "bcseqs is {'npbc1': 'AAGAAAGTTGTCGGTGTCTTTGT', 'npbc2': 'TCGATTCCGTTTGTAGTCGTCTG', 'npbc3': 'GAGTCTTGTGTCCCAGTTACCAG', 'npbc4': 'TTCGGATTCTATCGTGTTTCCCT', 'npbc5': 'CTTGTCCAGGGTTTGTGTAACCT', 'npbc6': 'TTCTCGCAAAGGCAGAAAGTAGT', 'npbc7': 'GTGTTACCGTGGGAATGAATCCT', 'npbc9': 'AACTAGGCACAGCGAGTCTTGGT', 'npbc10': 'AAGCGTTGAAACCTTTGTCCTCT', 'npbc8': 'TTCAGGGAACAAACCAAGTTACG', 'npbc11': 'GTTTCATCTATCGGAGGGAATGG', 'npbc12': 'CAGGTAGAAAGAAGCAGAATCGG'}\n",
      "condnames is ['11,[]_39,[sal]', '12.4,[]_37.6,[sal]', '13.2,[]_36.8,[sal]', '12,[atc]_38,[atc,sal]', '17.5,[atc]_32.5,[atc,sal]', '18.8,[atc]_31.2,[atc,sal]', '12.4,[atc,ara]_37.6,[atc,ara,sal]', '18,[atc,ara]_32,[atc,ara,sal]', '19.3,[atc,ara]_30.7,[atc,ara,sal]', '11.5,[atc,cin]_38.5,[atc,cin,sal]', '15.8,[atc,cin]_34.2,[atc,cin,sal]', '17.5,[atc,cin]_32.5,[atc,cin,sal]']\n",
      "put output into central\\groups\\murray-biocircuits\\ashur\\nanopore\\191114\\191114_simprec.py\n",
      "date of seq is 200131\n",
      "get reads from central\\groups\\murray-biocircuits\\ashur\\nanopore\\200131\\allreads.fastq\n",
      "prefixseqslist is ['AGGTATGATCCTGACGACGGAGCACGCCGTCGTCGACAAGCC']\n",
      "plasbcs is [['CCGGTTCGTTCCAAGCAACTAGTTTATCATCTAACTAG'], ['CTCACCAGCTGATGAGTAACTATTCCTGCTCCATTGGT', 'CTCACCAGCTGATGAGTAACTATTCCTGCTCCATTGGT']]\n",
      "postfixseq is ['CAAGCCCATTATTACCCTGTTATCCCTAGACACCAATCAGAGGCCACA']\n",
      "bcseqs is {'npbc7': 'GTGTTACCGTGGGAATGAATCCT', 'npbc1': 'AAGAAAGTTGTCGGTGTCTTTGT', 'npbc2': 'TCGATTCCGTTTGTAGTCGTCTG', 'npbc8': 'TTCAGGGAACAAACCAAGTTACG', 'npbc3': 'GAGTCTTGTGTCCCAGTTACCAG', 'npbc4': 'TTCGGATTCTATCGTGTTTCCCT', 'npbc9': 'AACTAGGCACAGCGAGTCTTGGT', 'npbc5': 'CTTGTCCAGGGTTTGTGTAACCT', 'npbc6': 'TTCTCGCAAAGGCAGAAAGTAGT', 'npbc10': 'AAGCGTTGAAACCTTTGTCCTCT', 'npbc11': 'GTTTCATCTATCGGAGGGAATGG', 'npbc12': 'CAGGTAGAAAGAAGCAGAATCGG'}\n",
      "condnames is ['5,[]_15,[sal]', '5,[]_15,[sal]', '5,[]_15,[sal]', '5,[atc]_15,[atc,sal]', '5,[atc]_15,[atc,sal]', '5,[atc]_15,[atc,sal]', '5,[atc,cin]_15,[atc,cin,sal]', '5,[atc,cin]_15,[atc,cin,sal]', '5,[atc,cin]_15,[atc,cin,sal]', '5,[atc,ara]_15,[atc,ara,sal]', '5,[atc,ara]_15,[atc,ara,sal]', '5,[atc,ara]_15,[atc,ara,sal]']\n",
      "put output into central\\groups\\murray-biocircuits\\ashur\\nanopore\\200131\\200131_simprec.py\n",
      "date of seq is 201125\n",
      "get reads from central\\groups\\murray-biocircuits\\ashur\\nanopore\\201125\\allreads.fastq\n",
      "prefixseqslist is ['TAATCTCATGACCAAAATCCCTTATGCCTAGGGATAACAGGGTAATGCCT', 'GATAAAAACGGTTAGCGCTTCGTTGCCTAGGGATAACAGGGTAATGCCTG']\n",
      "plasbcs is [['CTCACCAGCTGATGAGTAACTATTCCTGCTCCATTGGT', 'CCGGTTCGTTCCAAGCAACTAGTTTATCATCTAACTAG']]\n",
      "postfixseq is ['CAAGCCCATTATTACCCTGTTATCCCTAGACACCAATCAGAGGCCACA']\n",
      "bcseqs is {'npbc1': 'AAGAAAGTTGTCGGTGTCTTTGT', 'npbc2': 'TCGATTCCGTTTGTAGTCGTCTG', 'npbc3': 'GAGTCTTGTGTCCCAGTTACCAG', 'npbc4': 'TTCGGATTCTATCGTGTTTCCCT', 'npbc5': 'CTTGTCCAGGGTTTGTGTAACCT', 'npbc6': 'TTCTCGCAAAGGCAGAAAGTAGT', 'npbc7': 'GTGTTACCGTGGGAATGAATCCT', 'npbc8': 'TTCAGGGAACAAACCAAGTTACG', 'npbc9': 'AACTAGGCACAGCGAGTCTTGGT', 'npbc10': 'AAGCGTTGAAACCTTTGTCCTCT', 'npbc11': 'GTTTCATCTATCGGAGGGAATGG', 'npbc12': 'CAGGTAGAAAGAAGCAGAATCGG'}\n",
      "condnames is ['1,[sal,atc,cin]_22,[sal,atc,cin]', '1,[sal,atc,ara]_22,[sal,atc,ara]', '1,[sal]_22,[sal,atc,cin]', '1,[sal]_22,[sal,atc,ara]', '1,[sal,atc,ara]_22,[sal,atc,cin]', '1,[sal,atc,cin]_22,[sal,atc,ara]', '1,[sal,atc,cin]_22,[sal,atc,cin]', '1,[sal,atc,ara]_22,[sal,atc,ara]', '1,[sal]_22,[sal,atc,cin]', '1,[sal]_22,[sal,atc,ara]', '1,[sal,atc,ara]_22,[sal,atc,cin]', '1,[sal,atc,cin]_22,[sal,atc,ara]']\n",
      "put output into central\\groups\\murray-biocircuits\\ashur\\nanopore\\201125\\201125_simprec.py\n",
      "date of seq is 201212\n",
      "get reads from central\\groups\\murray-biocircuits\\ashur\\nanopore\\201212\\allreads.fastq\n",
      "prefixseqslist is ['TAATCTCATGACCAAAATCCCTTATGCCTAGGGATAACAGGGTAATGCCT', 'GATAAAAACGGTTAGCGCTTCGTTGCCTAGGGATAACAGGGTAATGCCTG']\n",
      "plasbcs is [['CTCACCAGCTGATGAGTAACTATTCCTGCTCCATTGGT', 'CCGGTTCGTTCCAAGCAACTAGTTTATCATCTAACTAG']]\n",
      "postfixseq is ['CAAGCCCATTATTACCCTGTTATCCCTAGACACCAATCAGAGGCCACA']\n",
      "bcseqs is {'npbc1': 'AAGAAAGTTGTCGGTGTCTTTGT', 'npbc2': 'TCGATTCCGTTTGTAGTCGTCTG', 'npbc3': 'GAGTCTTGTGTCCCAGTTACCAG', 'npbc4': 'TTCGGATTCTATCGTGTTTCCCT', 'npbc5': 'CTTGTCCAGGGTTTGTGTAACCT', 'npbc6': 'TTCTCGCAAAGGCAGAAAGTAGT', 'npbc7': 'GTGTTACCGTGGGAATGAATCCT', 'npbc8': 'TTCAGGGAACAAACCAAGTTACG', 'npbc9': 'AACTAGGCACAGCGAGTCTTGGT', 'npbc10': 'AAGCGTTGAAACCTTTGTCCTCT'}\n",
      "condnames is ['5,[sal]_25,[sal]', '5,[sal]_25,[sal,atc,cin]', '5,[sal,atc,ara]_25,[sal,atc,cin]', '5,[sal]_25,[sal,atc,ara]', '5,[sal,atc,cin]_25,[sal,atc,ara]', '5,[sal1]_25,[sal1]', '5,[sal1]_25,[sal1,atc,cin]', '5,[sal1,atc,ara]_25,[sal1,atc,cin]', '5,[sal1]_25,[sal1,atc,ara]', '5,[sal1,atc,cin]_25,[sal1,atc,ara]']\n",
      "put output into central\\groups\\murray-biocircuits\\ashur\\nanopore\\201212\\201212_simprec.py\n"
     ]
    }
   ],
   "source": [
    "seqlist = list(df_data.date_sequenced.unique())\n",
    "\n",
    "processreads = 1000\n",
    "frontchecklength = 175\n",
    "threshfrac = 0.3\n",
    "#bc_threshold = len(barcodes[\"BC01\"][0])*threshfrac\n",
    "#variable_sequence_threshold = len(bc1)*threshfrac\n",
    "#prefix_detection_threshold = len(prefix_seq)*threshfrac\n",
    "#end_threshold = len(postfix_seq)*threshfrac\n",
    "\n",
    "\n",
    "for seqdataset in seqlist:\n",
    "    \n",
    "    subdf = df_data[df_data.date_sequenced==seqdataset]\n",
    "    #prefix sequence(s)\n",
    "    prefixes = list(subdf.prefix.unique())\n",
    "    prefixseqslist = []\n",
    "    for prefix in prefixes:\n",
    "        if(pd.isna(prefix)):\n",
    "            prefixlist = []\n",
    "        elif(\",\" in prefix):\n",
    "            prefixlist = prefix.replace(\"[\",\"\").replace(\"]\",\"\").split(\",\")    \n",
    "        else:\n",
    "            prefixlist = [prefix]\n",
    "        for pref in prefixlist:\n",
    "            prefixseqslist += [df_seqs[df_seqs.name==pref].sequence.iloc[0]]\n",
    "    #barcode sequences\n",
    "    barcodes = list(subdf.barcode.unique())\n",
    "    bcseqs = {}\n",
    "    bcnames = []\n",
    "    condlist = []\n",
    "    for bc in barcodes:\n",
    "        bcseq = df_seqs[df_seqs.name==bc].sequence.iloc[0]\n",
    "        bcseqs[bc] = bcseq\n",
    "        #CONDITIONS\n",
    "        condstr = \"\"\n",
    "        for i in range(4):\n",
    "            cond = subdf[subdf.barcode==bc][\"c\"+str(i+1)].iloc[0]\n",
    "            if(pd.isna(cond)):\n",
    "                break\n",
    "            if(condstr != \"\"):\n",
    "                condstr += \"_\"+str(cond)\n",
    "            else:\n",
    "                condstr+=str(cond)\n",
    "        condlist += [condstr]\n",
    "    #plasmid barcodes\n",
    "    plasbcs = []\n",
    "    plasmidbarcodes = zip(list(subdf.variable1.unique()),list(subdf.variable2.unique()))\n",
    "    for plasbc1,plasbc2 in plasmidbarcodes:\n",
    "        if(pd.isna(plasbc1)):\n",
    "            plasbcs += [[]]\n",
    "        elif(pd.isna(plasbc2)):\n",
    "            plasbcs += [[df_seqs[df_seqs.name==plasbc1].sequence.iloc[0]]]\n",
    "        else:\n",
    "            plasbcs += [[df_seqs[df_seqs.name==plasbc1].sequence.iloc[0],\\\n",
    "                         df_seqs[df_seqs.name==plasbc2].sequence.iloc[0]]]\n",
    "    #postfix\n",
    "    postfix = []\n",
    "    postfixes = list(subdf.suffix.unique())\n",
    "    for pfix in postfixes:\n",
    "        pfixseq = df_seqs[df_seqs.name==pfix].sequence\n",
    "        if(len(pfixseq)==0):\n",
    "          postfix += [\"\"]\n",
    "        else:\n",
    "            postfix+=[pfixseq.iloc[0]]\n",
    "    if(any([pd.isna(bcseqs[a]) for a in bcseqs])):\n",
    "        continue\n",
    "    fastqfilename = os.path.join(datapath,str(seqdataset),readsname)\n",
    "    print(f\"date of seq is {seqdataset}\")\n",
    "    print(f\"get reads from {fastqfilename}\")\n",
    "    print(f\"prefixseqslist is {prefixseqslist}\")\n",
    "    print(f\"plasbcs is {plasbcs}\")\n",
    "    print(f\"postfixseq is {postfix}\")\n",
    "    print(f\"bcseqs is {bcseqs}\")\n",
    "    print(f\"condnames is {condlist}\")\n",
    "    print(f\"put output into {os.path.join(datapath,str(seqdataset),str(seqdataset)+'_'+str(outname))}\")\n",
    "    ''\n",
    "    allseqDict,seqstats,unsorted=barcodeSplitAndCountRecords(fastqfilename,bcseqs,\\\n",
    "                                                barcode_detection_threshold=len(bcseqs[0])*threshfrac,\\\n",
    "                                                end_threshold=len(postfix[0])*threshfrac,\\\n",
    "                                                processreads=processreads,\\\n",
    "                                                variable_sequences=plasbcs,\\\n",
    "                                                prefix_sequence=prefixseqslist,\\\n",
    "                                                postfix_sequence=postfix,\\\n",
    "                                               prefix_detection_threshold=len(prefixseqslist[0])*threshfrac,\\\n",
    "                                               variable_sequence_threshold=len(plasbcs[0])*threshfrac,\\\n",
    "                                                frontchecklength=frontchecklength,visualize=False,progressbar = False)\n",
    "    #'''\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ]
}